stown: new package contribution

rseichter opened the pull request:
New package "stown", new maintainer entry.

rseichter committed:
maintainers: add rseichter

In preparation for an upcoming new package contribution.

rseichter committed:
stown: init at 1.2.0

Manage file system object mapping via symlinks. Lightweight
alternative to GNU Stow.

Signed-off-by: Ralph Seichter <github@seichter.de>

SigmaSquadron commented on pkgs/by-name/st/stown/package.nix:
```diff
@@ -0,0 +1,46 @@
+{
+  fetchFromGitHub,
+  lib,
+  pkgs,
```
```suggestion
  tree,
```
Please call attributes directly in Nixpkgs. The `callPackage` function used internally will automatically find the correct attribute in the package set.

SigmaSquadron commented on pkgs/by-name/st/stown/package.nix:
```diff
@@ -0,0 +1,46 @@
+  fetchFromGitHub,
+  lib,
+  pkgs,
+  python3,
```
```suggestion
  python3Packages,
```
Usually, Python packages use `python3Packages` instead of `python3.pkgs`.

SigmaSquadron commented on pkgs/by-name/st/stown/package.nix:
```diff
@@ -0,0 +1,46 @@
+    "-v"
+  ];
+
+  meta = with lib; {
```
```suggestion
  meta = {
```
Wide `with` scopes are considered an [anti-pattern](https://nix.dev/guides/best-practices#with-scopes).

SigmaSquadron reviewed (commented):
Welcome to Nixpkgs! Please review the [commit conventions for `maintainers/`](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md); your maintainer commit should be named `maintainers: add rseichter`.

rseichter commented:
@SigmaSquadron Thank you for the review, that was mighty fast. I hope I addressed the points you mentioned in the proper way. My original package.nix was mostly a copy of an existing Python package build, and I opted to make as few changes as possible, seeing how this is my first ever Nix package contribution. Please let me know if anything else needs fixing.

SigmaSquadron reviewed (approved):
Looks good to me! The binary seems to build and run just fine.

---

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 389975`

### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>stown</li>
    <li>stown.dist</li>
  </ul>
</details>

rseichter commented:
@SigmaSquadron Just to make sure that I understand the process correctly: At this point I don't need to take any further action on my end, unless specifically told to do so, correct? In other words, I just wait for this pull request to be merged eventually?

SigmaSquadron commented:
That's right. Soon enough a committer will find this PR, conduct a final review, and merge it.

SigmaSquadron commented on pkgs/by-name/st/stown/package.nix:
```diff
@@ -0,0 +1,46 @@
+
+  nativeCheckInputs = [
+    python3Packages.unittestCheckHook
+    tree
+  ];
```
I don't think this is necessary, since `versionCheckHook` also checks for the version in `--help`'s output.

rseichter replied:
I have read [the docs](https://github.com/NixOS/nixpkgs/blob/master/doc/stdenv/stdenv.chapter.md) again, and since we have actual unit tests and you don't feel strongly about the subject, I think it is acceptable to rely on the default behaviour here.

rseichter replied:
Sorry, should be fixed with the latest Git push.

rseichter commented on pkgs/by-name/st/stown/package.nix:
```diff
@@ -0,0 +1,46 @@
+{
+  fetchFromGitHub,
+  lib,
+  python3Packages,
```
Hello @GaetanLepage . Does this mean you object to alphabetic ordering, or am I missing something? It should not have any semantic effect, or does it?

SigmaSquadron replied:
It's a mostly unwritten Nixpkgs convention that `lib` always comes first in the function call, followed by `stdenv` (In this case, the Python `stdenv`), followed by the fetcher. As mentioned, it's a nitpick.

rseichter replied:
Will fix.

rseichter commented on pkgs/by-name/st/stown/package.nix:
```diff
@@ -0,0 +1,46 @@
+  ];
+
+  nativeCheckInputs = [
+    python3Packages.unittestCheckHook
```
The upstream tests are written and intended for `unittest` exclusively. While they should theoretically work with `pytest`, that is neither guaranteed nor verified, so I don't feel comfortable introducing a dependency for the NixOS package alone.

rseichter replied:
Understood. I switched to pytest as you requested, with the caveat that if tests should fail at some point in the future, that constitutes what Douglas Adams classifies as Somebody Else's Problem. Others may prefer pytest, but I am planning to stick with Python's internal `unittest` framework for this package's upstream development. This is not meant to cause a fuzz, just for clarification.

rseichter commented on pkgs/by-name/st/stown/package.nix:
```diff
@@ -0,0 +1,46 @@
+    hash = "sha256-iHeqmlo7be28ISJfPZ7GZC2gj2VVgt20ORnfYVToo0A=";
+  };
+
+  nativeBuildInputs = [
```
Fixed.

rseichter commented on pkgs/by-name/st/stown/package.nix:
```diff
@@ -0,0 +1,46 @@
+
+  nativeBuildInputs = [
+    python3Packages.setuptools
+    python3Packages.wheel
```
Fixed.

rseichter commented on pkgs/by-name/st/stown/package.nix:
```diff
@@ -0,0 +1,46 @@
+    "-s"
+    "tests"
+    "-v"
+  ];
```
Removed.

GaetanLepage reviewed (approved):
Thanks @rseichter !

rseichter commented:
Thank you @GaetanLepage and @SigmaSquadron for your time helping me with this PR. Have a nice weekend.
